src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -797,7 +797,51 @@ private bool IsOpenable(bool needToUncompress, bool needToLoadIntoMemory, out st
             return true;
         }
 
-        private bool SizesTooLarge() => _compressedSize > uint.MaxValue || _uncompressedSize > uint.MaxValue;
+        private bool AreSizesTooLarge => _compressedSize > uint.MaxValue || _uncompressedSize > uint.MaxValue;
+
+        private static CompressionLevel MapCompressionLevel(BitFlagValues generalPurposeBitFlag, CompressionMethodValues compressionMethod)
+        {
+            // Information about the Deflate compression option is stored in bits 1 and 2 of the general purpose bit flags.
+            // If the compression method is not Deflate, the Deflate compression option is invalid - default to NoCompression.
+            if (compressionMethod == CompressionMethodValues.Deflate || compressionMethod == CompressionMethodValues.Deflate64)
+            {
+                return ((int)generalPurposeBitFlag & 0x6) switch
+                {
+                    0 => CompressionLevel.Optimal,
+                    2 => CompressionLevel.SmallestSize,
+                    4 => CompressionLevel.Fastest,
+                    6 => CompressionLevel.Fastest,
+                    _ => CompressionLevel.Optimal
+                };
+            }
+            else
+            {
+                return CompressionLevel.NoCompression;
+            }
+        }
+
+        private static BitFlagValues MapDeflateCompressionOption(BitFlagValues generalPurposeBitFlag, CompressionLevel compressionLevel, CompressionMethodValues compressionMethod)
+        {
+            ushort deflateCompressionOptions = (ushort)(
+                // The Deflate compression level is only valid if the compression method is actually Deflate (or Deflate64). If it's not, the
+                // value of the two bits is undefined and they should be zeroed out.
+                compressionMethod == CompressionMethodValues.Deflate || compressionMethod == CompressionMethodValues.Deflate64
+                    ? compressionLevel switch
+                    {
+                        CompressionLevel.Optimal => 0,
+                        CompressionLevel.SmallestSize => 2,
+                        CompressionLevel.Fastest => 6,
+                        CompressionLevel.NoCompression => 6,
+                        _ => 0
+                    }
+                    : 0);
+
+            return (BitFlagValues)(((int)generalPurposeBitFlag & ~0x6) | deflateCompressionOptions);
+        }
+
+        private bool IsOffsetTooLarge => _offsetOfLocalHeader > uint.MaxValue;
```

Comment:
Have you looked to make sure you'll always call this after the value has been set?  Is there an assert you could add to ensure that?

Reply:
Which value?

Reply:
No but I can take a look to see if we need to do any small refactoring in this area.

Reply:
I wonder if
```
 private bool IsOffsetTooLarge 
 {
    Debug.Assert(_offsetOfLocalHeader > 0);
    _offsetOfLocalHeader > uint.MaxValue;
}
```
Is enough, or if 0 is a valid value.

Reply:
Would you consider this suggestion a blocker for merging this servicing PR, or can I investigate this and fix it in a later change?

Reply:
I will definitely need more time to investigate this. I quickly tested this suggestion and I am seeing test failures. For now, I think we're ok with checking `IsOffsetTooLarge` even if the underlying field has not been updated. The main scenario that was failing has been confirmed to be fixed.